lib/redis_memo/memoizable/dependency.rb

```diff
@@ -40,6 +43,21 @@ def depends_on(dependency, **conditions)
     end
   end
 
+  def extract_dependencies_for_relation(relation)
+    # Extract the dependent memos of an Arel without calling exec_query to actually execute the query
+    RedisMemo::MemoizeQuery::CachedSelect.with_new_query_context do
+      connection = ActiveRecord::Base.connection
+      query, binds, _ = connection.send(:to_sql_and_binds, relation.arel)
+      RedisMemo::MemoizeQuery::CachedSelect.current_query = relation.arel
+      is_query_cached = RedisMemo::MemoizeQuery::CachedSelect.extract_bind_params(query)
+        raise(
+          RedisMemo::ArgumentError,
+          "Invalid Arel dependency. Query is not enabled for RedisMemo caching."
+        ) unless is_query_cached
+        extracted_dependency = connection.dependency_of(:exec_query, query, nil, binds)
```

Comment:
I think in this dependency_of, we will can still pick up the thread locals from the previous `extract_bind_params`.

I do like the `with_new_query_context` idea tho! Perhaps we should wrap dependency_of in that block instead.

Reply:
> Perhaps we should wrap dependency_of in that block instead.

Hm I think an issue is the memoized `exec_query` dependency block relies on the bind parameters being extracted previously in the overridden method [here](https://github.com/chanzuckerberg/redis-memo/blob/main/lib/redis_memo/memoize_query/cached_select/connection_adapter.rb#L14). I explicitly set the current_query thread local variable, and expect `extract_bind_params` to set the bind params thread local variable (retrieved in `CachedSelect.current_query_bind_params`).

I think we can get around this but would have to change the logic in the exec_query dependency block to be able to extract bind params on the fly. 

> I think in this dependency_of, we will can still pick up the thread locals from the previous extract_bind_params.

Hm not sure I understand this. Unless another db query occurs within the block that `with_new_query_context` is wrapped around, we should be okay right?

Reply:
Oh I guess 

> Unless another db query occurs within the block that `with_new_query_context` is wrapped around, we should be okay right?

Oh yeah I guess that's true. We don't call exec_query in exec_query.

The case I was thinking about is to have
```ruby
memoize_method :site_names do |teacher|
  teacher.sites.each do |site|
    depends_on Site.where(id: site.id)
  end
end
```
but the inner exaction would not be affected by the outer query anyway.